### ISSUE TITLE
`source.unusedImports` is very slow and adds unnecessary whitespace

### ISSUE THREAD

**Issue by torext** [microsoft/pylance-release#3715]
<h3> Environment data </h3>

-   Language Server version: 2022.11.40
-   OS and version: linux x64
-   Python version (and distribution if applicable, e.g. Anaconda): 3.10.8, Anaconda
-   python.analysis.indexing: false
-   python.analysis.typeCheckingMode: off

<h3> Repro Steps </h3>

1. Open a `.py` file with some unused imports and some code code below with proper formatting (see e.g. example GIFs below).
2. Run `Fix All` from the command palette, making sure the following settings are enabled:
```
    "python.analysis.fixAll": [
        "source.unusedImports",
    ]
```

<h3> Expected behavior </h3>

With version 2022.11.30 the unused imports would be removed almost instantaneously and no additional whitespace would be added to the file.

<h3> Actual behavior </h3>

The unused imports take around 3 seconds to get removed and unnecessary whitespace is added at the bottom of the file. The below two GIFs show the behaviour for the current and the previous release.

Current (2022.11.40):
![Peek 2022-12-05 11-34](https://user-images.githubusercontent.com/10012123/205619803-9ac1c18c-b18f-4e51-9ff6-1ab90a00157e.gif)

Previous (2022.11.30):
![Peek 2022-12-05 11-35](https://user-images.githubusercontent.com/10012123/205619910-0cad1c03-603c-4d31-8823-cc371561ea32.gif)

**Comment by torext** [comment]
Here's a minimal example showcasing the same issue, note the presence of newline (\n) as a last character:
```
from math import factorial

import numpy as np
import pandas as pd


def f(n):
    return factorial(n)

```
After running `Fix All`, which takes roughly 3-4 seconds agains, the result is the following. Note the extra blank line at the end and between the import and the function definition:
```
from math import factorial



def f(n):
    return factorial(n)


```

**Comment by rchiodo** [comment]
The whitespace I believe is by design. It doesn't eliminate the whitespace below the change, it simply sends a T[extEdit](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textEdit) that says remove the unused imports.

VS code then decides how to format the file after the edits are applied.

**Comment by torext** [comment]
> The whitespace I believe is by design. It doesn't eliminate the whitespace below the change, it simply sends a [TextEdit](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textEdit) that says remove the unused imports.

I think you misread my posts: I'm not talking about the whitespace between the imports and the function, I'm talking about an extra blank line being added at the very bottom of the file. Version 2022.11.30 didn't use to do this. Let me show you again with the simpler example (you'll also see the very noticeable speed difference again):

2022.11.40:
![Peek 2022-12-07 09-50](https://user-images.githubusercontent.com/10012123/206132980-1939b287-7643-435b-8502-f2e2e627781f.gif)

2022.11.30:
![Peek 2022-12-07 09-52](https://user-images.githubusercontent.com/10012123/206133019-38abbc51-bb93-4e13-9446-93352d0b69b7.gif)

So you see the current version doesn't just delete the lines containing unused imports, it adds an additional blank line at the very end of the file; the old version didn't use to do this. I'm convinced this can't be a design choice and even if it is, it's a horrible one...

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2022.12.21, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20221221-7-December-2022-PreRelease)
